Adding map ex

Frenchris opened the pull request.

kigiri commented on exercise/map.js:
```diff
@@ -0,0 +1,56 @@
+let arrRandom1 = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
+
+for (let i = 0; i < 10; i++) {
+  let value = Math.round(Math.random() * (10 - -10 - 10))
```
🥇 `(10 - -10 - 10)` most complicated way of writting `10` I have seen yet.

kigiri commented on exercise/map.js:
```diff
@@ -0,0 +1,56 @@
+
+for (let i = 0; i < 10; i++) {
+  let value = Math.round(Math.random() * (10 - -10 - 10))
+  arrRandom1.push(value)
```
If you need random values in your tests, it's best not to expose them outside of tests.
For that you can use test context:

```js
t((_, ctx) => ctx.arrRandom1 = Array(10).fill().map(() => Math.round(Math.random() * 10)))
```
you can then use your context like so:
```js
t(({ eq }, ctx) => eq(fn(ctx.arrRandom1)[0], check(arr[0]))
```

Frenchris replied:
Noted. Declaring them in the setup now.

kigiri commented on exercise/map.js:
```diff
@@ -0,0 +1,56 @@
+https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map
+
+*/
+let arr1 = [10, -10, 20, -95, 86]
```
this array is so small they can just write the solution doing the math in theyr head.
You don't even need JS for that haha

Frenchris replied:
I wanted to show an easy "example" one along with a random one.

kigiri commented on exercise/map.js:
```diff
@@ -0,0 +1,56 @@
+2) Then the results have to be subtracted by 4
+3) finally, the negative values have to be replaced by null
+then do the same with arrRandom1 by returning arrRandom2
+The use of `for` and `while` is forbidden for this exercise
```
I think this should be a functions, and even more that we should have a few different use cases.
Not just 1 use of map.

I also don't thing this is a very demonstrative usecase of map.

Frenchris replied:
reworked the subject to ask for functions.

kigiri commented on exercise/map.js:
```diff
@@ -0,0 +1,56 @@
+})
+
+t(({ code }) => !code.includes('for'))
+t(({ code }) => !code.includes('while'))
```
I think a better way to ensure they used map would be to "monkey patch" map:

```js
t((_, ctx) => {
  ctx.mapCall = []
  const { map } = Array.prototype
  Array.prototype.map = function () {
    ctx.mapCalls.push([this, ...arguments])
    return map.apply(this, arguments)
  }
})
```
now this wierd shitty code allow to wrap the *classic* `[].map` in a function that save the arguments of each calls.

so in your tests you can check in the context if they used map or not, and try to mimic the call:
```js
t(({ eq }, ctx) => eq(fn([10, -2, 4]), [16, null, 0]) // 1st test execute it
t(({ eq }, ctx) => eq(ctx.mapCalls[ctx.mapCalls.length - 1][0], [10, -2, 4]) // check that map was called on this array
```

Frenchris replied:
In use now. I transmitted the knowledge to Afonso.

kigiri reviewed (commented):
not fund of `map1`, `map2` and `map3` names

kigiri reviewed (changes requested):
while having `choumi` arround is fun, I dislike including us in the subject, I'd like them to be more neutral.

If we think about when to use map:
- Change the shape of an array (ex: *(from array of user `object` to array of name `string`)*)
	```js
	[{ name: 'vanessa' }, { name: 'gaetan' }].map(user => user.name)
	```

- Apply a computations on an array of numbers
	this is where you did `map1` but really what's the point of your calculation, it's wierd and confusing.
	```js
	[0, 10.10, 87, 55, 80].map(fahrenheitToCelsius)
	```
	something like that achieve the same goal and is actually relatable.
    I mean I could see how this is usefull.

- Changing only some elements in an array:
	```js
	const changePassword = (users, id, password) =>
		users.map(user => user.id === id ? {...user, password } : user)
	```

The problem I have with your exemples is that they feel far from data we would have in a program or realife usecase.

please make sure to add tests with `Object.freeze` so we confirm that they didn't edit the array :)

Frenchris commented:
Did an update of the subject based on the last two comments. Awaiting feedback.

Frenchris committed:
1)  changed order of functions asked
2) reformatted with ``` ```js around the code examples

Frenchris committed:
1) rename map into "using map" to adopt the new naming standard
2) removed afonso outdated solution
3) removed the bumbering in the subject
4) pushed the data down below the tests

kigiri commented on exercise/using-map.js:
````diff
@@ -0,0 +1,353 @@
+    temperature: ' 84 ° F   ',
+  },
+]
+```js
````
it's:
\`\`\`js
\`\`\`
not:
\`\`\`
\`\`\`js

kigiri commented on exercise/using-map.js:
```diff
@@ -0,0 +1,353 @@
+The function returns then an array of strings.
+
+Example:
+['alabama', 'new jersey']
```
no js here ?

kigiri commented on exercise/using-map.js:
```diff
@@ -0,0 +1,353 @@
+    city: 'San Francisco',
+    temperature: '84°F',
+  },
+]
```
inline those:
```js
[
  { city: 'Los Angeles', temperature: '101°F' },
  { city: 'San Francisco', temperature: '84°F' },
]
```
that's plenty readable.

kigiri commented on exercise/using-map.js:
```diff
@@ -0,0 +1,353 @@
+
+https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map
+
+*/
```
`*/` is extraneous

kigiri commented on exercise/using-map.js:
```diff
@@ -0,0 +1,353 @@
+
+- Map
+
+https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map
```
only links, and from devdocs if possible.

kigiri commented on module/call-me-maybe.js:
```diff
@@ -1,9 +1,20 @@
+  // curry: { feedback: true },
+  // curry2: { feedback: true },
+  // 'do-your-own-curry': { feedback: true },
+  // some: { feedback: true },
```
rebase with gh-pages and remove the comments durring the rebase conflicts.

Frenchris committed:
fixed formatting issues

Frenchris committed:
fix of using-map.js (lost a line during conflict resolvement)

Frenchris committed:
updated solution for 'using-map' (former map) for frenchris

Frenchris committed:
fix of formatting
